Compiles on Ubuntu 10.04 with Qt 4.6

henrycg opened the pull request:
- One minor change (removing QSet<QSharedPointer>)
- All tests passed
- Things seem to run fine

henrycg commented:
Hi David,

This morning I merged again with your code and ran the tests. The TCP disconnect segfault problem is no longer there (thanks!). The diff is massive. Most of the changes involve Tunnel and the Tolerant rounds.

If you want to look through the diff and come back to me with comments, that works too.

Henry

davidiw commented:
Sorry just saw this now, I guess I'm not getting e-mail notifications!  I'll begin looking through it.

Cheers,
David

davidiw commented on src/Anonymity/BulkRound.cpp:
```diff
@@ -64,6 +64,10 @@
 
   void Xor(QByteArray &dst, const QByteArray &t1, const QByteArray &t2)
   {
+    if(t1.count() != t2.count()) {
```
This can be dangerous if you don't make it explicit the the t1 and t2 have to be the same length, if you look below, we take the min size of the inputs and output.

henrycg replied:
I just added a comment in the docstring for the method. Since XORing two bitstrings of unequal length is always an error (right?) I think we should fail loudly when it happens. If not, we could add a return value to the function and check it every time.

davidiw commented on src/Anonymity/Session.cpp:
```diff
@@ -27,8 +27,7 @@
```
I was thinking about this, and I think you are on to the right idea, but more concretely, we need to implement some form of datagramming service on top of the session, this could be a nice addition to the paper for making practical systems in anonymity systems.  What you've got started is in the right direction, but we certainly don't want to be throwing away packets somewhat silently.

henrycg replied:
There are a lot of different ways to do this: the application can handle datagramming, we can add a new layer between the session and application, etc. My preference is to make as minimalistic of an API as possible inside of the session. If we need a full-fledged datagramming stack, we can add it in a separate layer.

davidiw commented on src/Anonymity/Session.hpp:
```diff
@@ -148,7 +148,7 @@
 
       static const int MinimumRoundSize = 3;
 
-      static const int PeerJoinDelay = 10000;
```
why the delta?

davidiw commented on src/ClientServer/CSConnectionAcquirer.cpp:
```diff
@@ -41,7 +41,7 @@
   {
     TimerCallback *cb = new Utils::TimerMethod<CSConnectionAcquirer, int>(
         this, &CSConnectionAcquirer::RequestServerState, -1);
-    _check_event = new TimerEvent(Timer::GetInstance().QueueCallback(cb, 120000, 120000));
```
Why the frequency change?

davidiw commented on src/Crypto/AsymmetricKey.cpp:
```diff
@@ -3,7 +3,7 @@
 
 namespace Dissent {
 namespace Crypto {
-  int AsymmetricKey::DefaultKeySize = 512;
```
For tests, I think we should leave this at 512, maybe we should have build flags ... debug vs release....

davidiw commented on src/Tests/RoundTest.cpp:
```diff
@@ -41,7 +41,7 @@
   {
     Timer::GetInstance().UseVirtualTime();
 
```
Why the delta?

henrycg replied:
That was just a mistake

davidiw commented:
Left some comments.  Your tunnel code looks good (from skimming), I appreciate the clarity and comments!

Commit:
Working on session dynamics

Commit:
Anonymity working

Commit:
Got dynamics to work

Commit:
Adding back command line flags for emulab

Commit:
Adding disconnect handler

Commit:
Adding debug

Commit:
Fixing error message

Commit:
Adding check for offline peer

Commit:
Reverting server group size

Commit:
Removing line that breaks relay

Commit:
Adding server commit to final message

Commit:
Replacing XOR length check

Commit:
Adding debugging messages

Commit:
Removing warnings and adding debug

Commit:
Trying to track down bug with RNGS

Commit:
Adding debug

Commit:
more debug

Commit:
Adding debug

Commit:
Adding potential RNG fix'

Commit:
Fix

Commit:
Removing extra debug

Commit:
Fix to timeout code for very very slow servers

Commit:
Merge

Commit:
Merge with CSBulk

Commit:
Adding utils for UCB web trace replays

Commit:
Fix tunnel bug

Commit:
Adding tunnel debug code

Commit:
Fix map/hash bug

Commit:
Adding makefile

Commit:
Increasing slot length

Commit:
Utils for ucb trace testing through proxy

Commit:
Fixing script for proxy urls

Commit:
Updating time script

Commit:
Adding interrupt code in client

Commit:
Adding outgoing proxy to tunnel

Commit:
Removing failure responses

Commit:
Fixing bug

Commit:
Fixing cleartext size bug

Commit:
Adding fix to proxy for tunnel

Commit:
Merge with davidiw

Commit:
Keeping slots open once opened

Commit:
Bug fix to length

Commit:
Client flushes output after every request

Commit:
Adding top1000 sites script

Commit:
Merge branch 'davidiw' of github.com:henrycg/Dissent into davidiw

Commit:
Updating run script

Commit:
Updating run script'

Commit:
switching to use mulk for multithreaded downloads

Commit:
Fixes to run.py script

Commit:
Adding BlogDrop parameter generation

Commit:
Adding BlogDrop plaintext methods

Commit:
Blogdrop public and private keys

Commit:
Adding modular inverse

Commit:
Adding server ciphertext functions

Commit:
Adding first client cipehrtext proof

Commit:
Client ciphertext proofs working

Commit:
Merge remote-tracking branch 'davidiw/master' into blogdrop

Commit:
Adding check that elements are in subgroup

Commit:
Saving intermediate work on blogdrop round

Commit:
Fixing bugs in blogdrop round

Commit:
Renaming 'Modulus' operator to correct name 'Modulo'

Commit:
Efficiency and bug fixes for BlogDropRound

- Using PowCascade to speed up exponentiation
- Servers can send messages
- Passes Round tests
- Use smaller keys for testing
- Removed inapplicable tests

Commit:
Fixing doxygen comments

Commit:
Adding test that includes non-authors

Commit:
Adding BlogDrop wrappers

- BlogDropClient, BlogDropAuthor, BlogDropServer
- End-to-End test

Commit:
Merge with david

Commit:
Changing author API to reject long plaintext

Commit:
BlogDrop bins use QSharedPointer to save space

Commit:
Changing to QSharedPointers for all BlogDrop objs

Commit:
Merge remote-tracking branch 'davidiw/master' into blogdrop

Commit:
BlogDropRound uses bin containers

Commit:
Parameters allow for ciphertexts of diff lengths

Commit:
Ciphertexts can be of different lengths

Commit:
Making Integer destructor virtual

Commit:
Fixing bug in SessionFactory that kept BlogDrop round from working

Commit:
Command-line options do not overwrite .conf file

Commit:
Fix settings error

Commit:
Boosting timeouts for BlogDrop

Commit:
Fix bug in Plaintext

Commit:
Fix to test case

Commit:
Servers can use per-client threading

Commit:
Removing unnecessary checks to speed up verification

Commit:
Adding 0x prefix to remove ambiguity

Commit:
Adding benchmarking test for platform comparison

Commit:
Using Params::Testing instead of ifdef's to avoid mistakes

Commit:
Saving preliminary work on abstract groups

Commit:
Fixing bugs pointed out by David

Commit:
Abstracting away Group so that we can use EC with BlogDrop

Commit:
Adding documentation to abstract group code

Commit:
Saving work on EC in BlogDrop

- Code compiles and tests run w/o errors
- Working on first-cut integration of EC groups
- Not sure if we should make separate abstraction (as in
    Integer) for ECPoint and ECCurve

Commit:
Adding fatal note so I dont forget to fix proto

Commit:
Adding fatal note so I dont forget to fix proto

Commit:
ECGroup mostly works

Commit:
Wrote test suite for BlogDrop over EC

Commit:
Merge remote-tracking branch 'davidiw/master' into blogdrop

Commit:
BlogDropRound updates

- Fix commit bug pointed out by David
- Addressing David's misc code comments

Commit:
BlogDropRound with EC passes tests

Commit:
Reverting to one proof per ciphertext

Having many ClientCiphertext elements in each slot
was much too confusing and inefficient. Reverting to
old design.

Commit:
Variable-length proofs work

Commit:
Adding tool for generating prime groups

Commit:
Removing binary

Commit:
Adding gitignore

Commit:
Adding 2048 integer group

Commit:
Adding EC benchmarks

Commit:
BlogDrop work

- Adding benchmark test of crypto++ Integer vs ECP
- Adding ability to prove knowledge of secret key

Commit:
Servers prove knowledge of secrets

Commit:
Merge with David's latest code

Commit:
Disabling threading: need to add synchronization for it to be safe

Commit:
Saving work on pairings

Commit:
BlogDrop passes all tests -- proof abstraction works

Commit:
Adding util to generate pbc params

Commit:
Saving work on pairing... still not working

Commit:
Pairing group works

Commit:
Pairing group in GT works also

Commit:
Updating todo

Commit:
Integer conversion uses long-lived temp

Commit:
Adding PBCWrapper ext

Commit:
Moving EC -> CppEC in preparation for adding OpenSSL EC

Commit:
Saving work on openssl ec

Commit:
Closer to functioning with OpenSSL

Commit:
update todo

Commit:
Merge branch 'blogdrop' of github.com:henrycg/Dissent into blogdrop

Commit:
BlogDrop works with OpenSSL

Commit:
BlogDrop OpenSSL EC group works

Commit:
Pairing wrapper works

Commit:
Updating TODO

Commit:
Working on PBC GT encode/decode

Commit:
Trying to map into GT

Commit:
Working on encoding elements in GT

Commit:
Can encode in elements of GT

Commit:
Adding proof comments

Commit:
Aesthetic changes

Commit:
Getting ready to add new types of proofs

Commit:
Blogdrop round works

Commit:
Saving work on pairing ciphertext

Commit:
Todo

Commit:
Reorg of pairing code

Commit:
First-cut pairing proof works

Commit:
Fixed encoding into GT

Commit:
BlogDrop pairing works start to finish

Commit:
Testing out OpenSSL integers

Commit:
Adding code to test if element is a generator

Commit:
Got hashing generator group functions working

Commit:
Abstracting changing generator proofs

Commit:
Part way through hashing gen construction

Commit:
TODO

Commit:
Todo

Commit:
Hashing abstraction works

Commit:
Working on OpenSSL integer integration

Commit:
Fixing bug in OpenSSL integers

Commit:
Adding skeleton for benchmarking

Commit:
gitignore

Commit:
First round of benchmarking code

Commit:
Bench code and gitignore

adboomlodestar committed:
Utils for benchmarking

adboomlodestar committed:
Working on benchmarks

adboomlodestar committed:
Saving code

adboomlodestar committed:
BlogDropRound works

adboomlodestar committed:
Using compressed encoding of GT

Commit:
Working on blogdroproudn with hashing

Commit:
BlogDropRound works with hashing

Commit:
XOR benchmark code for blogdrop

Commit:
XOR benchmark working

Commit:
todo

Commit:
Merge
